### ISSUE TITLE
master/v5.0.x: mpirun --hostfile broken

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#10156]
On master, `mpirun --hostfile NAME ...` is broken:

```
$ mpirun --version
mpirun (Open MPI) 5.1.0a1

Report bugs to https://www.open-mpi.org/community/help/
$ git rev-parse HEAD
eb87378cd8c55c08fde666c03a04aa2811da395a
$ cat h
mpi009
mpi010
$ mpirun --hostfile h -np 2 hostname
Provide a hostfile
```

Same thing happens at HEAD of the v5.0.x branch.

There are two issues:

1. The file clearly exists and was provided on the command line.
2. The help message is terrible -- it needs to be more clear.

**Comment by awlauria** [comment]
You actually hit an edge-case. It works with hostfiles not named 'h' as one would expect.

prtre has reserved the right to use 'h' as an alias to give the help output for a specific option. For example:

```
$ ./exports/bin/prterun --host h
--------------------------------------------------------------------------
Comma-separated list of hosts to invoke processes on
--------------------------------------------------------------------------
```

prrte seems to want to support this:

```
$ ./exports/bin/prterun --help h
Help can be requested in three ways:

- a directive to the cmd. For example, the command
    "prterun --help" or "prterun -h"
  will provide the high-level help output giving a general
  description of each available option.

- an option to the help directive. For example, the command
    "prterun --help foo" or "prterun -h foo"
  will provide detailed help about the "foo" command line
  directive.

- specifying "help" as an option to the directive of interest.
  For example, the command
    "prterun --foo --help" or "prterun --foo -h"
  will provide detailed help about the "foo" command line
  directive. Note that the "help" and "h" options do not
  require the dash prefixes - i.e., the following commands
    "prterun --foo help" or "prterun --foo h"
  will also output detailed help for "foo". Thus, directives
  are precluded from defining options of "help" and "h".

Note that the last two methods are functionally equivalent
and will return the same output.
```

Notice the last bit: 

**Note that the "help" and "h" options do not
  require the dash prefixes - i.e., the following commands
    "prterun --foo help" or "prterun --foo h"
  will also output detailed help for "foo". Thus, directives
  are precluded from defining options of "help" and "h".**

I think we can work around this in ompi by passing in a flag or checking that we are using the mpi command line.

Honestly I don't like the support of non-dashed 'h' and 'help' in this case, seems ripe for edge cases. But it is what it is and I don't know if that is something we want to push back on.

**Comment by jsquyres** [comment]
That is terrible.  `h` is a valid filename.

The token argument to `--hostfile` should never be interpreted as anything other than a filename.

My $0.02: if there's some additional logic going on that might end up treating the argument to `--hostfile` -- or any valid CLI option -- as something else, then that logic should be fixed.

**Comment by bwbarrett** [comment]
+1.  `h` is, for better or for worse, a commonly used example hostfile name in web pages.  We can't break that.

**Comment by awlauria** [comment]
I agree.

I think the most sane thing is to remove support for the non-dashed help options rather than trying to work-around it - it's a bug not specific to the ompi/schizo as it also impacts prterun/prun.

Allowing "prterun --hostfile -h" or "prterun --hosefile --help" seems sane and good enough - I don't get the rationale of trying to support the non-dashed versions. It just leads to headaches, because of course hostfile/machinefile aren't the only options impacted. A quick scan/test shows any --long-option 'h' will be.

Removing it is trivial, and in the case openpmix pushes back on that,  it's also easy enough to work-around it via an env or similar.

**Comment by jsquyres** [comment]
@awlauria Ok.  I quibble with one thing, though.  I think:

```
pretrun --hostfile -h ...
```

should look for a file named `-h`.  That `-h` should _not_ be interpreted as a "please show the help message" option.

Specifically: the argument to `--hostfile` should _always_ be interpreted as a filename.  This is how many other CLI parsers (including `getopt_long(3)`) work.

**Comment by awlauria** [comment]
I don't think I would go that far.

`./exports/bin/prterun --hostfile "\-h" hostname` works on current prte master for users who want to name their hostfiles/paths in such a manner. Which is about what I would expect.

"-h" is short for --help. Even trying to open a file names "-h" with `vi` takes some effort (`vi "\-h"` is how I did it.)

fwiw: trying `mpirun --hostfile -h` doesn't work on a v4 flavor I am on, but `mpirun --hostfile "\-h"` does.

**Comment by jsquyres** [comment]
I'm not sure if you're agreeing with me or not.  😄 

I agree that `-h` is a kinda stupid filename.  But the argument to `--hostfile` should always, _Always,_ **ALWAYS** be interpreted as a filename.  If someone types in `mpirun --hostname -h uptime`, it should open the file named `-h` and use that (and error if there is no file named `-h`).

I'm not going to advocate for what `prterun` should do.  I'm only talking about `mpirun` (I know they're sorta the same thing, but I assumed that there was a different schitzo component involved in handling argv[0] == `prterun` vs. argv[0] == `mpirun`).

**Comment by bwbarrett** [comment]
@awlauria vi takes some effort because the filename isn't an argument of an option, but the "leftover" arguments.  If you do `vi -- -h`, you'll edit a file named `-h` as you'd expect.  But take an application that takes a filename as an option argument, like sed:

```
% cat -- -h
s/foo/FOO/g
% echo "foobar" | sed -f -h 
FOObar
```

Yes, that's pretty contrived, but I was actually struggling to find an easy example while typing this.  So I think that `prterun --hostfile -h` should, according to getopt_long behavior, "just work" using a file named `-h` as its hostfile.  You can also see this behavior by running the example in the Linux `getopt_long` man page.  For example:

```
% ./getopt_example --add "add_string"
option add with arg add_string
% ./getopt_example --delete "delete_string"
option delete with arg delete_string
% ./getopt_example --add --delete    
option add with arg --delete
```

**Comment by awlauria** [comment]
Ah, I see. 

So should we just drop this prte feature from mpi tools?

**Comment by awlauria** [comment]
I'm referencing that prte allows users to do:

`mpirun --option -h/--help/help h`

to get a show help message referencing `--option`.

As I mentioned, this bug is not limited to `--hostfile`, but could impact other options (I was able to reproduce with `--path` in a similar manner).

**Comment by jsquyres** [comment]
IMHO, that feature is not good CLI design because the tokens `-h`, `--help`, `help`, and `h` can be valid arguments in many cases.

I don't know the history or rationale of this design, but it feels like this feature may have been inspired by CLIs like `git`, where you can do things like `git log --help`.  But -- at least in the `git` scenario -- that's a little different because the git `log` command (and many others) doesn't take a mandatory argument.  Hence, the use of `--help` is unambiguous.  In mpirun's case, `--hostfile` has a mandatory argument, so there's always going to be ambiguity in the values of the mandatory argument: is a filename or is it a modifier to `--hostfile`?

It would be better to not have the ambiguity at all.

**Comment by bosilca** [comment]
Maybe we do want do follow some [guidelines](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html).

**Comment by gpaulsen** [comment]
This will be fixed on `main` when submodule pointers are updated, presumably by https://github.com/open-mpi/ompi/pull/10165.

**Comment by awlauria** [comment]
This was included in v5.0.x here: https://github.com/open-mpi/ompi/pull/10179
